File: java/src/main/java/benchmarks/CalculatorAst.java
package benchmarks;

import java.util.*;

public class CalculatorAst extends Benchmark {

    interface Node {}

    static class Number implements Node {
        final long value;
        Number(long value) {
            this.value = value;
        }
    }

    static class Variable implements Node {
        final String name;
        Variable(String name) {
            this.name = name;
        }
    }

    static class BinaryOp implements Node {
        final char op;
        final Node left;
        final Node right;
        BinaryOp(char op, Node left, Node right) {
            this.op = op;
            this.left = left;
            this.right = right;
        }
    }

    static class Assignment implements Node {
        final String variable;
        final Node expr;
        Assignment(String variable, Node expr) {
            this.variable = variable;
            this.expr = expr;
        }
    }

    private long resultVal;
    private String text;
    public List<Node> expressions = new ArrayList<>();
    public long n;

    public CalculatorAst() {
        n = configVal("operations");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Calculator::Ast";
    }

    private String generateRandomProgram(long lines) {
        StringBuilder sb = new StringBuilder();
        sb.append("v0 = 1\n");

        for (int i = 0; i < 10; i++) {
            int v = i + 1;
            sb.append("v").append(v).append(" = v").append(v - 1)
            .append(" + ").append(v).append("\n");
        }

        for (long i = 0; i < lines; i++) {
            int v = (int)(i + 10);
            sb.append("v").append(v).append(" = v").append(v - 1).append(" + ");

            switch (Helper.nextInt(10)) {
            case 0:
                sb.append("(v").append(v - 1).append(" / 3) * 4 - ").append(i)
                .append(" / (3 + (18 - v").append(v - 2).append(")) % v")
                .append(v - 3).append(" + 2 * ((9 - v").append(v - 6)
                .append(") * (v").append(v - 5).append(" + 7))");
                break;
            case 1:
                sb.append("v").append(v - 1).append(" + (v").append(v - 2)
                .append(" + v").append(v - 3).append(") * v").append(v - 4)
                .append(" - (v").append(v - 5).append(" /  v").append(v - 6).append(")");
                break;
            case 2:
                sb.append("(3789 - (((v").append(v - 7).append(")))) + 1");
                break;
            case 3:
                sb.append("4/2 * (1-3) + v").append(v - 9).append("/v").append(v - 5);
                break;
            case 4:
                sb.append("1+2+3+4+5+6+v").append(v - 1);
                break;
            case 5:
                sb.append("(99999 / v").append(v - 3).append(")");
                break;
            case 6:
                sb.append("0 + 0 - v").append(v - 8);
                break;
            case 7:
                sb.append("((((((((((v").append(v - 6).append(")))))))))) * 2");
                break;
            case 8:
                sb.append(i).append(" * (v").append(v - 1).append("%6)%7");
                break;
            case 9:
                sb.append("(1)/(0-v").append(v - 5).append(") + (v").append(v - 7).append(")");
                break;
            }
            sb.append("\n");
        }

        return sb.toString();
    }

    @Override
    public void prepare() {
        text = generateRandomProgram(n);
    }

    static class Parser {
        private final String input;
        private int pos;
        private final char[] chars;
        final List<Node> expressions = new ArrayList<>();

        Parser(String input) {
            this.input = input;
            this.chars = input.toCharArray();
        }

        void parse() {
            while (pos < chars.length) {
                expressions.add(parseExpression());
            }
        }

        private Node parseExpression() {
            Node node = parseTerm();

            while (pos < chars.length) {
                skipWhitespace();
                if (pos >= chars.length) break;

                char ch = chars[pos];
                if (ch == '+' || ch == '-') {
                    char op = ch;
                    advance();
                    Node right = parseTerm();
                    node = new BinaryOp(op, node, right);
                } else {
                    break;
                }
            }

            return node;
        }

        private Node parseTerm() {
            Node node = parseFactor();

            while (pos < chars.length) {
                skipWhitespace();
                if (pos >= chars.length) break;

                char ch = chars[pos];
                if (ch == '*' || ch == '/' || ch == '%') {
                    char op = ch;
                    advance();
                    Node right = parseFactor();
                    node = new BinaryOp(op, node, right);
                } else {
                    break;
                }
            }

            return node;
        }

        private Node parseFactor() {
            skipWhitespace();
            if (pos >= chars.length) return new Number(0);

            char ch = chars[pos];
            if (ch >= '0' && ch <= '9') {
                return parseNumber();
            } else if (ch >= 'a' && ch <= 'z') {
                return parseVariable();
            } else if (ch == '(') {
                advance();
                Node node = parseExpression();
                skipWhitespace();
                if (pos < chars.length && chars[pos] == ')') {
                    advance();
                }
                return node;
            } else {
                return new Number(0);
            }
        }

        private Node parseNumber() {
            long value = 0;
            while (pos < chars.length && chars[pos] >= '0' && chars[pos] <= '9') {
                value = value * 10 + (chars[pos] - '0');
                advance();
            }
            return new Number(value);
        }

        private Node parseVariable() {
            int start = pos;
            while (pos < chars.length &&
                    ((chars[pos] >= 'a' && chars[pos] <= 'z') ||
                     (chars[pos] >= '0' && chars[pos] <= '9'))) {
                advance();
            }
            String varName = input.substring(start, pos);

            skipWhitespace();
            if (pos < chars.length && chars[pos] == '=') {
                advance();
                Node expr = parseExpression();
                return new Assignment(varName, expr);
            }

            return new Variable(varName);
        }

        private void advance() {
            if (pos < chars.length) pos++;
        }

        private void skipWhitespace() {
            while (pos < chars.length && Character.isWhitespace(chars[pos])) {
                advance();
            }
        }
    }

    @Override
    public void run(int iterationId) {
        Parser parser = new Parser(text);
        parser.parse();
        expressions = parser.expressions;
        resultVal += expressions.size();

        if (!expressions.isEmpty() && expressions.get(expressions.size() - 1) instanceof Assignment) {
            Assignment assign = (Assignment) expressions.get(expressions.size() - 1);
            resultVal += Helper.checksum(assign.variable);
        }
    }

    @Override
    public long checksum() {
        return resultVal;
    }

    public List<Node> getExpressions() {
        return expressions;
    }
}

File: kotlin/src/main/kotlin/benchmarks/CalculatorAst.kt
package benchmarks

import Benchmark

class CalculatorAst : Benchmark() {
    sealed class Node

    data class Number(
        val value: Long,
    ) : Node()

    data class Variable(
        val name: String,
    ) : Node()

    data class BinaryOp(
        val op: Char,
        val left: Node,
        val right: Node,
    ) : Node()

    data class Assignment(
        val variable: String,
        val expr: Node,
    ) : Node()

    var n: Long = 0
    private var resultVal: UInt = 0u
    private lateinit var text: String
    lateinit var expressions: List<Node>

    init {
        n = configVal("operations")
    }

    private fun generateRandomProgram(lines: Long = 1000): String =
        buildString {
            append("v0 = 1\n")
            for (i in 0 until 10) {
                val v = i + 1
                append("v$v = v${v - 1} + $v\n")
            }
            for (i in 0 until lines) {
                val v = i + 10
                append("v$v = v${v - 1} + ")
                when (Helper.nextInt(10)) {
                    0 -> append("(v${v - 1} / 3) * 4 - $i / (3 + (18 - v${v - 2})) % v${v - 3} + 2 * ((9 - v${v - 6}) * (v${v - 5} + 7))")
                    1 -> append("v${v - 1} + (v${v - 2} + v${v - 3}) * v${v - 4} - (v${v - 5} /  v${v - 6})")
                    2 -> append("(3789 - (((v${v - 7})))) + 1")
                    3 -> append("4/2 * (1-3) + v${v - 9}/v${v - 5}")
                    4 -> append("1+2+3+4+5+6+v${v - 1}")
                    5 -> append("(99999 / v${v - 3})")
                    6 -> append("0 + 0 - v${v - 8}")
                    7 -> append("((((((((((v${v - 6})))))))))) * 2")
                    8 -> append("$i * (v${v - 1}%6)%7")
                    9 -> append("(1)/(0-v${v - 5}) + (v${v - 7})")
                }
                append("\n")
            }
        }

    override fun prepare() {
        text = generateRandomProgram(n)
    }

    private class Parser(
        private val input: String,
    ) {
        private var pos = 0
        private val chars = input.toCharArray()
        private val length = chars.size
        val expressions = mutableListOf<Node>()

        fun parse(): List<Node> {
            while (pos < length) {
                expressions.add(parseExpression())
            }
            return expressions
        }

        private fun parseExpression(): Node {
            var node = parseTerm()

            while (pos < length) {
                skipWhitespace()
                if (pos >= length) break

                val ch = currentChar()
                if (ch == '+' || ch == '-') {
                    val op = ch
                    advance()
                    val right = parseTerm()
                    node = BinaryOp(op, node, right)
                } else {
                    break
                }
            }

            return node
        }

        private fun parseTerm(): Node {
            var node = parseFactor()

            while (pos < length) {
                skipWhitespace()
                if (pos >= length) break

                val ch = currentChar()
                if (ch == '*' || ch == '/' || ch == '%') {
                    val op = ch
                    advance()
                    val right = parseFactor()
                    node = BinaryOp(op, node, right)
                } else {
                    break
                }
            }

            return node
        }

        private fun parseFactor(): Node {
            skipWhitespace()
            if (pos >= length) return Number(0)

            return when (val ch = currentChar()) {
                in '0'..'9' -> {
                    parseNumber()
                }

                in 'a'..'z' -> {
                    parseVariable()
                }

                '(' -> {
                    advance()
                    val node = parseExpression()
                    skipWhitespace()
                    if (currentChar() == ')') {
                        advance()
                    }
                    node
                }

                else -> {
                    Number(0)
                }
            }
        }

        private fun parseNumber(): Node {
            var value = 0L
            while (pos < length && chars[pos].isDigit()) {
                value = value * 10 + (chars[pos] - '0')
                advance()
            }
            return Number(value)
        }

        private fun parseVariable(): Node {
            val start = pos
            while (pos < length && (chars[pos].isLetterOrDigit())) {
                advance()
            }
            val varName = input.substring(start, pos)

            skipWhitespace()
            if (pos < length && currentChar() == '=') {
                advance()
                val expr = parseExpression()
                return Assignment(varName, expr)
            }

            return Variable(varName)
        }

        private fun currentChar(): Char = if (pos < length) chars[pos] else '\u0000'

        private fun advance() {
            if (pos < length) pos++
        }

        private fun skipWhitespace() {
            while (pos < length && chars[pos].isWhitespace()) {
                advance()
            }
        }
    }

    override fun run(iterationId: Int) {
        val parser = Parser(text)
        expressions = parser.parse()
        resultVal += expressions.size.toUInt()
        if (expressions.isNotEmpty() && expressions.last() is Assignment) {
            val assign = expressions.last() as Assignment
            resultVal += Helper.checksum(assign.variable)
        }
    }

    override fun checksum(): UInt = resultVal

    override fun name(): String = "CalculatorAst"
}
